### PR TITLE
rate-limiting: add global rate limiting APIs

### PR DESCRIPTION
**Pull request by shashankram**
**Description**:
Adds the global rate limiting capability to the existing
rate limiting APIs based on the [global rate limiting design](https://docs.google.com/document/d/1fNfDvOV0p93DxP0vLGJmP4FTqd4ZgggI9PnA-sZWnWI/edit?usp=sharing).

- Supports TCP connection rate limiting
- Supports HTTP request rate limiting at the VirtualHost
  and Route level.

Part of #2018

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [X] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? `no`
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change? `no`

3. Has documentation corresponding to this change been updated in the [osm-docs](https://github.com/openservicemesh/osm-docs) repo (if applicable)? `not yet`

### CODE REVIEW

**Review by nojnhuh — APPROVED** [review]

**Comment by keithmattix** [comment]
I'm planning to review next week after reading the design doc

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go [hunk 1]
```diff
@@ -193,6 +200,226 @@ type HTTPLocalRateLimitSpec struct {
...
+// HTTPGlobalRateLimitDescriptor defines rate limit descriptor to use
+// in the rate limit service request for HTTP requests.
+type HTTPGlobalRateLimitDescriptor struct {
+	// Entries defines the list of rate limit descriptor entries.
+	Entries []HTTPGlobalRateLimitDescriptorEntry `json:"entries,omitempty"`
+}
+
+// HTTPGlobalRateLimitDescriptorEntry defines the rate limit descriptor entry
+// to use in the rate limit service request for HTTP requests.
+type HTTPGlobalRateLimitDescriptorEntry struct {
```
Comment (steeling) [inline pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:319]: are these intended to act as a `oneof` ?

Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:320]: Yes, updated the comment to remove ambiguity.

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go [hunk 2]
```diff
@@ -193,6 +200,227 @@ type HTTPLocalRateLimitSpec struct {
...
+	RemoteAddress *RemoteAddressDescriptorEntry `json:"remoteAddress,omitempty"`
+
+	// RequestHeader defines a descriptor entry that is generated only when the
+	// request header matches the given header name. The value of the descriptor
+	// entry is derived from the value of the header present in the request.
+	// +optional
+	RequestHeader *RequestHeaderDescriptorEntry `json:"requestHeader,omitempty"`
+
+	// HeaderValueMatch defines a descriptor entry that is generated when the
+	// request header matches the given HTTP header match criteria.
```
Comment (keithmattix) [inline pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:337]: nit: missing +optional here?

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go [hunk 3]
```diff
@@ -193,6 +200,227 @@ type HTTPLocalRateLimitSpec struct {
...
+	Regex string `json:"regex,omitempty"`
+
+	// Contains defines the substring value to match.
+	// +optional
+	Contains string `json:"contains,omitempty"`
+
+	// Present defines whether the request matches the criteria
+	// when the header is present. If set to false, header match
+	// will be performed based on whether the header is absent.
+	Present *bool `json:"present,omitempty"`
```
Comment (keithmattix) [inline pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:421]: Is there a default?

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go [hunk 4]
```diff
@@ -193,6 +200,227 @@ type HTTPLocalRateLimitSpec struct {
...
+	// Value defines the value of the descriptor entry.
+	Value string `json:"value"`
+}
+
+// HTTPGlobalRateLimitSpec defines the global rate limiting specification
+// for HTTP requests.
+type HTTPGlobalRateLimitSpec struct {
+	// RateLimitService defines the rate limiting service to use
+	// as a global rate limiter.
+	RateLimitService RateLimitServiceSpec `json:"rateLimitService"`
```
Comment (keithmattix) [inline pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:279]: Clarifying question: what's the difference in convention here between a pointer type vs. a value type? I see a mix here (e.g. the fields of GlobalRateLimitSpec)

**Review by keithmattix — APPROVED** [review]
LGTM - just getting some clarifications

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go (hunk 4 above)
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:279]: Per API convention, pointer types are used for optional fields.

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go (hunk 2 above)
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:337]: Good catch, fixed.

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go (hunk 3 above)
Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:422]: No. One of these must be set explicitly to either true/false.

File: pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go (hunk 4 above)
Comment (keithmattix) [reply pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:279]: But the bool referenced in my other comment is a pointer, yet it's not optional. I'm guessing it's to deal with the zero value being false?

Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:279]: Could you specify which field are you referring to?

Comment (keithmattix) [reply pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:279]:
```go
// Present defines whether the request matches the criteria
// when the header is present. If set to false, header match
// will be performed based on whether the header is absent.

Present *bool `json:"present,omitempty"`
```

Comment (shashankram) [reply pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go:279]: Added the `+optional` tag to the `Present` field.

**Pull request merged**